qos/evm/state_archival.go

```diff
@@ -0,0 +1,190 @@
+
+	// TODO_IMPROVE(@commoddity): set an expiry time for the archival state so that a new random block can be selected on an interval.
+}
+
+func (a *archivalState) isEnabled() bool {
+	return a.archivalCheckConfig.Enabled
+}
+
+// processEndpointArchivalData processes all archival-related data from an endpoint
+// and updates the archival state accordingly.
+//
+// Returns true if the endpoint's archival data was processed successfully,
+// false if the endpoint should be skipped (e.g., missing archival balance).
+func (a *archivalState) processEndpointArchivalData(endpoint endpoint) bool {
+	if !a.isEnabled() {
+		return true
+	}
+
+	// Attempt to retrieve the archival balance from the endpoint
+	balance, err := endpoint.getArchivalBalance()
+	if err != nil {
+		return false
+	}
+
+	// Update the consensus map to determine the balance at the perceived block number
+	a.updateConsensusMap(balance)
+	return true
+}
+
+// updateConsensusMap updates the balance consensus map to determine
+// the expected archival balance at the archival block number.
+//
+// Once <archivalConsensusThreshold> endpoints report the same balance,
+// the balance is set as the expected archival balance.
+func (a *archivalState) updateConsensusMap(balance string) {
+	if a.expectedBalance == "" && balance != "" {
+		a.balanceConsensus[balance]++
+	}
+}
+
+// updateArchivalState updates the archival state based on the perceived block number.
+// This handles calculating the archival block number if needed and updating the
+// consensus-based archival balance.
+func (a *archivalState) updateArchivalState(perceivedBlockNumber uint64) {
+	// Skip further processing if archival checks are not enabled
+	if !a.isEnabled() {
+		return
+	}
+
+	// If the archival block number is not yet set for the service, calculate it.
+	// This requires that the perceived block number is set to determine the latest possible block number.
+	if perceivedBlockNumber != 0 && a.blockNumberHex == "" {
+		a.logger.Info().Msg("Calculating archival block number")
+		a.calculateArchivalBlockNumber(perceivedBlockNumber)
+	}
+
+	// If the expected archival balance is not yet set for the service, set it.
+	// This utilizes the consensus map to determine a source of truth for the archival balance.
+	// If <archivalConsensusThreshold> endpoints report the same balance, it is considered the source of truth.
+	if a.expectedBalance == "" {
+		a.updateArchivalBalance(archivalConsensusThreshold)
+	}
+}
+
+// calculateArchivalBlockNumber determines a random archival block number based on the perceived block number.
+// The function applies the following logic:
+//   - If perceived block is below threshold, returns block 0
+//   - Otherwise, calculates a random block between minArchivalBlock and (perceivedBlockNumber - threshold)
+//   - Ensures the returned block number is never below the contract start block
+func (a *archivalState) calculateArchivalBlockNumber(perceivedBlockNumber uint64) string {
```

Comment:
Overall, moving most of the code to this file is a great improvement.
Another iteration (or a TODO_IMPROVE) can organize everything into a single, clear flow of data/events.

Reply:
It's been largely refactored.